py/desispec/zcatalog.py

```diff
@@ -1,15 +1,429 @@
+    tab = vstack(tables)
+    
+    ## Selecting primary spectra for the whole combined ZCATALOG
+    ## For SV, it selects the best spectrum including cmx+special+sv1+sv2+sv3
+    ## For Main, it selects the best spectrum for main+special
+    nspec, specprim = find_primary_spectra(tab)
+    
+    ## Replacing the existing 'ZCAT_NSPEC' and 'ZCAT_PRIMARY'
+    ## If all_columns = False, and user-list does not contain this column -
+    ## these columns will be added 
+    tab['ZCAT_NSPEC'] = nspec
+    tab['ZCAT_PRIMARY'] = specprim
+    
+    ## Adding empty columns for SV|MAIN NSPEC and PRIMARY
+    col1 = Column(np.array([0]*len(tab)), name = f'{survey_name.upper()}_NSPEC', dtype = '>i8')
+    col2 = Column(np.array([0]*len(tab)), name = f'{survey_name.upper()}_PRIMARY', dtype = 'bool')
+    tab.add_columns([col1, col2], [-4, -3])
+    
+    ## Selecting primary spectra for targets within just SV or MAIN depending on the survey_name
+    ## For SV, it selects the primary spectra out of SV1+SV2+SV3 for every individual TARGETID
+    ## Ignores cmx+special in SV
+    ## For Main, it selects the primary spectra just for 'main' and ignores 'special'
+    is_survey = np.char.startswith(tab['SURVEY'].astype(str).data, survey_name)
+    nspec, specprim = find_primary_spectra(tab[is_survey])
+    tab[f'{survey_name.upper()}_NSPEC'][is_survey] = nspec
+    tab[f'{survey_name.upper()}_PRIMARY'][is_survey] = specprim
+    
+    ## For convenience, sort by SURVEY, PROGRAM, and (HEALPIX or TILEID) 
+    if (version == 'zpix'):
+        tab.sort(['SURVEY', 'PROGRAM', 'HEALPIX'])
+    else:
+        tab.sort(['SURVEY', 'PROGRAM', 'TILEID', 'LASTNIGHT'])
+    
+    ## Add the final table as a HDU
+    hdu0 = fits.PrimaryHDU()
+    hdu1 = fits.BinTableHDU(tab)
+    
+    ## Final HDUList
+    hdu = fits.HDUList([hdu0, hdu1])
+    ## Save this to the output_filename location
+    hdu.writeto(output_filename, overwrite = True)
+    
 ####################################################################################################
 ####################################################################################################
 
+def add_target_columns(tab, survey, program):
+    """
+    The function adds extra '*_TARGET' columns that are not available in a given redshift catalog.
+    This is needed only for SV data, given the different TARGET columns for each SURVEY.
+    Adding empty columns is needed before stacking tables with astropy to avoid MaskedColumns.
+    
+    Parameters
+    ----------
+    tab : Astropy Table
+        The redshift catalog which needs the columns to be added.
+    
+    survey : str
+        The SURVEY of the given redshift catalog (cmx|special|sv1|sv2|sv3)
+        
+    program : str
+        The PROGRAM of the given redshift catalog (dark|bright|backup|other)
+        
+    Returns
+    -------
+    tab : Astropy Table
+        The final table after adding all the required '*_TARGET' columns
```

Comment:
Please clarify in the docstring whether this modifies the input `tab` in-place, or whether it leaves the input unchanged and returns the new table with additional columns.  If modifying the input table in-place, consider *not* returning the table, which sort of makes it more obvious that it is modifying the table in-place.  Context: That's one of the reasons that `list.sort()` doesn't return the sorted copy -- they want to you realize that it is modifying the list in place so that you don't accidentally think that `sorted_list = orig_list.sort()` gives you a new list without changing the original.